msticpy/analysis/polling_detection.py

```diff
@@ -0,0 +1,170 @@
+        Detection in Noisy Biological Time Series," 2007 IEEE International Workshop on Genomic
+        Signal Processing and Statistics, 2007, pp. 1-4, doi: 10.1109/GENSIPS.2007.4365817.
+        [2] https://github.com/cran/GeneCycle/blob/master/R/fisher.g.test.R
+
+        """
+        if exclude_pi:
+            pxx = pxx[:-1]
+
+        pxx_length = len(pxx)
+        test_statistic = np.max(pxx) / sum(pxx)
+        upper = np.floor(1 / test_statistic).astype("int")
+
+        if pxx_length > 700:
+            p_value = 1 - (1 - np.exp(-pxx_length * test_statistic)) ** pxx_length
+        else:
+            compose = []
+            for j in range(1, upper):
+                compose.append(
+                    (-1) ** (j - 1)
+                    * np.exp(
+                        np.log(special.binom(pxx_length, j))
+                        + (pxx_length - 1) * np.log(1 - j * test_statistic)
+                    )
+                )
+
+            p_value = sum(compose)
+
+        p_value = min(p_value, 1)
+
+        return test_statistic, p_value
+
+    def detect_polling(
+        self,
+        timestamps: npt.NDArray,
+        process_start: int,
+        process_end: int,
+        interval: int = 1,
+    ) -> float:
+        """
+        Carry out periodogram polling detecton.
+
+        Carries out the the procedure outlined in [1] to detect if the arrival times have a strong
+        periodic component.
+        The procedure estimates the periodogram for the data and passes the results to fishers G
+        test.
+
+        For more information run PeriodogramPollingDetector._g_test.__doc__
+
+        This code was adapted from [2].
+
+        Parameters
+        ----------
+        timestamps: ArrayLike
+            An array like object containing connection arrival times as timestamps
+        process_start: int
+            The timestamp representing the start of the counting process
+        process_end: int
+            The timestamp representing the end of the counting process
+        interval: int
+            The interval in seconds between observations
+
+        Returns
+        -------
+        p_val: float
+            The p value from fishers G test
+
+        References
+        ----------
+          [1] Heard, N. A. and Rubin-Delanchy, P. T. G. and Lawson, D. J. (2014) Filtering
+          automated polling traffic in computer network flow data. In proceedings of IEEE
+          Joint Intelligence and Security Informatics Conference 2014
+          [2] https://github.com/fraspass/human_activity/blob/master/fourier.py
+
+        """
+        time_steps = np.arange(process_start, process_end, step=interval)
+        counting_process = Counter(timestamps)
+
+        dn_ = np.array([counting_process[t] for t in time_steps])
+        dn_star = dn_ - len(timestamps) / len(time_steps)
+
+        freq, pxx = signal.periodogram(dn_star)
+
+        max_pxx_freq = freq[np.argmax(pxx)]
+
+        print(
```

Comment:
Not sure if printing multiple lines per detection pass might be very noisy (esp if you have lots of edges).
You could return this as part of the result - maybe a tuple or dict?
If called from the DF function I sketched out above you could return the whole thing as a DF with cols:
- concated group ID
- p_value
- start
- end
- dominant interval
- frequency

Reply:
Yes, that makes a lot of sense, I don't think I thought this one through!

I'll be implementing the DF idea above so the `detect_polling_arr` function will return a tuple of **p_value, dominant_frequency, dominant_interval**. 
The columns for these 3 values returned will then be created in the DF

Reply:
Sounds good!